### PR TITLE
Zoom Bar

### PR DESCRIPTION
**Pull request by Keelando**
This branch is mistakenly named Draggable lattice, which will come immediately after this PR.

WIP on the zoom functionality. There are still some bugs to work out but I wanted to have a mock-up of how it would work. Unfortunately at the moment its a little laggy, and will only get worse with larger lattices.

Have a look and send some suggestions.

### CODE REVIEW

**Review by gwwatkin — COMMENTED** [review]
Zoom feature is really cool. ~~Do you plan on adding panning?~~ I see that's draggable lattice

Not laggy on my computer but there is some stickyness in the zoom bar. Probably a matter of how refreshes happen. Might be caused by having `defaultValue` and `value` in the `Slider`, you probably need to remove `defaultValue`.

Watch out, the dashed borders are gone ;)

**Comment by gwwatkin** [comment]
Thanks for sharing :slightly_smiling_face:

**Comment by Keelando** [comment]
I can try removing default. I suspect it's running multiple onChange events as you slide it, which is somewhat mitigated by allowing only discrete step intervals (customizable). Theres an onChangeEnd prop, which waits til the slider stops to run, but we then lose control of the 'controlled' slider component, ie, since the input value is tied to the component state, it stops refreshing. Just some weird react logic to figure out.

**Comment by Keelando** [comment]
Fixed the sticky slider @gwwatkin

**Comment by Keelando** [comment]
Two more features that are coming:

- fixed position of zoom bar/slice selector when looking at huge lattices so its always visible
- draggable lattices

Should I integrate those in this PR? or do you want to have a reivew?

**Comment by Keelando** [comment]
> Watch out, the dashed borders are gone ;)

I didn't understand this comment. Did something break?

**Comment by gwwatkin** [comment]
> I can try removing default. I suspect it's running multiple onChange events as you slide it, which is somewhat mitigated by allowing only discrete step intervals (customizable). Theres an onChangeEnd prop, which waits til the slider stops moving to fire, but we then lose control of the 'controlled' slider component, ie, since the input value is tied to the component state, it stops refreshing. Just some weird react logic to figure out.

Now you the slider doesn't update live. Removing the default not work? usually having defaultValue and value in react components is not desireable

**Comment by gwwatkin** [comment]
> Also, the behaviour gets a bit weird at the bottom end of the zoom scale, so I can either fix that edge case (reduce content displayed in each cell because it won't be legible anyway) or just truncate the zoom bar so it doesn't go to 0%

You can make the text disappear below a certain threshold. But however you think is best. I think having live updates work should take priority because if you change the slider and the lattice blows up in your face is a bit confusing :sweat_smile:

**Comment by gwwatkin** [comment]
> I didn't understand this comment. Did something break?
How it is in this PR
Screenshot_20220301_150810

How it should be:
Screenshot_20220301_151001
These boundaries are the main feature of lattice surgery. The JSON contains the info about which is rough/dashed and which is smooth/solid.

**Review by gwwatkin — COMMENTED** [review]
I think once the boundary problem is fixed we can merge. 

I do think it would be beneficial to have the zoom bar be interactive if it's not a lot of work (i.e. no more than just removing the default value and restoring the on change).

**Comment by Keelando** [comment]
> > I didn't understand this comment. Did something break?
> > How it is in this PR
> > Screenshot_20220301_150810
> 
> How it should be: Screenshot_20220301_151001 These boundaries are the main feature of lattice surgery. The JSON contains the info about which is rough/dashed and which is smooth/solid.

Ahh I get you now. Unfortunately this mistake appears to have crept on my earlier PR.. will sort it out. I think the border css is getting overwritten/superceded somehwere

**Comment by gwwatkin** [comment]
> Unfortunately this mistake appears to have crept on my earlier PR

Dev is fine. I think it's in this PR

**Comment by Keelando** [comment]
Dev is fine because I didn't push the last PR to production, whoops. So the problem was trying to offload some inline CSS to a class, and it was overwriting the inline css that is generating the borders etc in cell viewer. There's probably a way to extend css rather than overwrite, but thats some nitty gritty css stuff.

**Comment by Keelando** [comment]
> I think once the boundary problem is fixed we can merge.
> 
> I do think it would be beneficial to have the zoom bar be interactive if it's not a lot of work (i.e. no more than just removing the default value and restoring the on change).

I think the "stickyness" previously encountered in the slider is a performance issue due to multiple lattice re-renders. Each time the lattice is rendered, we have to loop through the lattice and set the cell-dimensions for each cell. I might be able to set the cell-dimension as a css attribute on a class and just change the class. I'll see what I can do. I think that rendering the zoom on the onChangeEnd, as it is in this PR is likely the most performant solution

File: src/components/CellViewer.tsx [hunk 1]
```diff
@@ -45,21 +44,32 @@ const styles_map_activity_color: StylesMapActivityColorType = {
...
 }
 
-const CellViewer = ({ cell, row_idx, col_idx }: CellViewerProps) => {
+const CellViewer = ({ cell, cell_font_size, row_idx, col_idx }: CellViewerProps) => {
+    const cellFontSize = (cell: VisualArrayCell) =>
+        cell.text.length > 20
+            ? cell_font_size
+            : cell.text.length > 7
+            ? `${cell_font_size * 1.3}`
+            : `${cell_font_size * 2}`
```
Comment (gwwatkin) [inline src/components/CellViewer.tsx:60]: Note for a future PR. If you know of a better way to do this go ahead and change it. I know this is how I had it set up

File: src/components/Zoombar.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                min={0}
+                max={sliderMax}
+                step={sliderStep}
+                onChange={(val) => {
+                    setSliderValue(val)
+                    setSliderMoving(true)
+                }}
+                onChangeEnd={(val) => {
+                    setCellDimension(val)
+                    setCellFontSize(val / 7.5)
```
Comment (gwwatkin) [inline src/components/Zoombar.tsx:72]: General comment, no change requested. Is there a particular reason why we have separate state for `fontSize` and `cellDimension`? you could have only `cellDimension` and replace and define

```ts
const cellFontSize = () => cellDimension/7.5
```

**Review by gwwatkin — APPROVED** [review]
> I could include logic to let the lattice size update in real time from the slider when lattices are small, then switch to waiting until onChangeEnd to render large lattices? 

Probably not worth it. If `onChange` makes the UI difficult to use on large lattices we don't use it. Not too big of a difference

Nice touch with the shadow

File: src/components/Zoombar.tsx (hunk 2 above)
Comment (Keelando) [reply src/components/Zoombar.tsx:72]: Nice, I think I will clean that up. It was a happy accident that the font size was a nice ratio of the cell dimension, lol

Comment (Keelando) [reply src/components/Zoombar.tsx]: That actually improved performance a little bit maybe.

**Pull request merged**